React Native retro v3 (#547)
artsy/artsy.github.io: _posts/2019-03-17-three-years-of-react-native.markdown

---
layout: epic
title: "React Native, 3 years later"
date: "2019-03-17"
author: [orta]
categories: [reactnative, ios, community, roads and bridges]
series: React Native at Artsy
comment_id: 521
---

On Valentine's day in 2014, @alloy made our first commit moving the Artsy Mobile team to JavaScript, and paving the

> way to the shared Omakase JS stack across web + iOS. We've done a write-up at [6 months], [1 year], [2 years] and

Reviewer: ```suggestion
way to the [shared Omakase JS stack across web + iOS][oma]. We've done a write-up at [6 months], [1 year], [2 years] and
```

at 2.5 years we collaborated on a React Native conference with Facebook which features a [very long Q&A][q_a]
session with the people who worked on, and with our React Native stack.

Our experience has been really positive building a single platform data-driven app. We've been able to drastically

> increase the number of contributors to the codebase and we barely need to train web-developers to have them be

Reviewer: > we barely need to train web-developers

I'd maybe rephrase this as:

> with minimal guidance, web-developers are able to be productive and ship features to our iOS apps.

productive in shipping features to our iOS apps.

That said, for this 3 year anniversary, I want to dive deep into some of the less positive aspects aspects of our
transition. We think these trade-offs are worth it, and that this may be what a successful cultural transition
eventually looks like for some companies.

<!-- more -->

## De-nativification

When adopting React Native, we the de-empathized iOS as a unique platform for Artsy.

From an engineering team's perspective, we think of it as skill [de-siloing][desilo]. Prior to the move, if you
were on the mobile team you only worked on the iOS apps. This meant you had a limited scope to make change at
Artsy. This comes from two factors:

> - Skills in Obj-C and Swift are only useful in the context of a single platform.

Reviewer: ```suggestion
- Skills in Obj-C and Swift are only useful in the context of Apple's platforms.
```

- Internally and externally, Artsy is considered a website

> Within 2 years we had de-siloed mobile engineering completely. We started with a team of 5 experts in iOS native

Author: better

and by the end everyone had very solid skills across the board in React, Node, GraphQL and the build tools we'd

> need to make it all come together. These engineers kept their native skills, but they were blunted with less need

Author: Yeah, maybe there's a better way to describe this

> need to make it all come together. These engineers kept their native skills, but they were blunted with less need

Reviewer: This is how I read it, though I agree with it. There are useful new UIKit APIs that we miss and don't get implemented – here's an example: https://github.com/artsy/eigen/issues/2787

> need to make it all come together. These engineers kept their native skills, but they were blunted with less need

Author: Maybe frozen is the right metaphor?

for them.

With time, we redistributed the native engineers across many teams, with the native Engineers effectively acting as
a conduit for ensuring that we keep quality high and providing guidance to the rest of that team on how to make it
feel right. A simple way to think of it, is that the native engineer's job was to make sure we still conformed to
the Apple Human Interface Guideline.

We're reasonably lucky here too, the mobile team at Artsy has pre-dominantly hired folks interested in improving
[Roads and Bridges][rnb] style infrastructure. With the move we effectively took all of our native product/feature
developers and moved them into infrastructure roles for JavaScript developers.

This was one of the major blocking points for AirBnB's adoption of React Native, their native teams felt

> uncomfortable at the introduction of JavaScript and a whole new toolchain inside their apps. There are people who

Author: Agree, no, nothing I can think of that's public (there were deleted tweets when I researched)

really love being a product developer in a native codebase and React Native as a technology de-values that, because
React can make it easy to go from data to screen so much better.

(Note: this is a gross simplification, you should read from the horses mouth and check out [their posts][airbnb]
(and [Ash's continuation][ash_airbnb] on the subject ))

After a year of adopting React Native, we used to joke that no-one enjoyed working in the native codebase anymore.

> They still did. Three years down the line, with most screens now in React Native, that's not a joke anymore.

Author: Former, I'm not sure that we really have new native iOS devs

> They still did. Three years down the line, with most screens now in React Native, that's not a joke anymore.

Reviewer 2: Yeah. I meant “iOS” devs, i.e. those that use RN 😄

This has consequences.

## Platform Concerns

De-siloing our mobile team obviously wasn't without its risk. In expanding the scope of our mobile engineering
team, and opening up the iOS app for contributions to the rest of the company we:

- Moved the engineers with a native focus to act more like platform engineers vs product engineers
- Had iOS native platform engineers with the skills to also work on the web's platform

This is what started to make our third year tricky. In Artsy, web is the [squeaky wheel][wheel].

> On the web side, we saw the most amount of growth in hiring and complexity needed to handle that. After figuring

Author: More the necessary changes in complexity

front-end at Artsy. It wasn't surprising when the aspect with the most company focus became much larger. What
creates tricky constraints are that the few people with the interest and skills to work on our React Native
infrastructure also happen to have the skills to work on the web problems too.

Those web problems tend to be a lot more valuable to the business.

There's some work that can be done to benefit both, but in the last year it has been hard to prioritise
iOS-specific platform work. For example, it took almost a year to get around to upgrading our version of React
Native. This is a pretty risky place to be for a platform which we care about.

At the end of 2018, we came to the conclusion that this was something we wanted to work to improve. So, we
specifically brought back an explicit iOS-specific culture to Artsy - when we split our front-end practice into web
and [iOS][fe-ios].

Maybe this is a small admission of failure to the idea of a purely de-silo'd team, but realistically while product
work across the two front-ends teams is consistent - the platform concerns just aren't.

> ## Keeping up with the Jobs'

Reviewer: I think this was intentional, as in `keeping up with the Jones'`, an English idiom about comparing oneself to their neighbours.

> ## Keeping up with the Jobs'

Author: Nah, this one is useful - Jobs' is many jobs (as in Steve Jobs)

> ## Keeping up with the Jobs'

Reviewer 2: Ohhhh, I see. Makes sense!

WWDCs and iOS releases just kinda happen, and we keep moving independently. We still support iOS 9, and have very
little incentive to bump it - new APIs from Apple just aren't that interesting anymore.

This is a side-effect to the openness of the JavaScript toolchain, and our ability to contribute and fix a lot of
our own problems at JavaScript level instead of needing to dive deeper into the native side to work on problems.

Without a focused iOS team, it's not easy to pitch for iOS specific projects. This means that features like moving
to Apple's new app store review API or using iOS features which don't exist on web gets talked about in meetings,
but never started. This lack of a focused team makes it really hard to implement fixes to bad App Store reviews

That said, in the last year we did manage to ship a pretty hefty [ARKit feature][arkit] - which a positive example
of an iOS-specific feature which is both forward-thinking tech and a super great fit for Artsy's product. React
Native basically played no part in that.

## Community Disconnection

There are maybe four communities at play if you're doing React Native:

- Node (Babel/TypeScript/Storybooks/VSCode)

> - React (Hooks/Relay/Styled Components)

Author: Eh, just something that's kinda fresh in mind, it's kinda interesting to know what folks are doing with it but maybe it doesn't fit with the rest tho

- React Native (Mostly people creating cross-platform tools)
- Native (Obj-C/Swift/CocoaPods/Fastlane etc)

It's hard to keep on-top of any one community, and it's very hard to keep on track of four. Realistically, if you

> want to be writing apps at the level of quality we do - you need to though.

Reviewer 2: ```suggestion
want to be writing apps at the level of quality we want to - you need to, though.
```

> Being involved in so many communities, as a part of owning [the Omakase][oma] stack means that it's hard to engage

Reviewer 2: I’m unsure what the “as a part of owning the Omakase stack” section is meant to say here. I feel like the sentence needs reworking.

in the native community with too much excitement anymore. Most of their problems aren't the same as ours.

## Universal Issues?

Are these the kind of problems most teams would have? It depends, with React Native at Artsy our focus on:

> - Offering a consistent way to write code across web and iOS, which didn't water down either platform

Reviewer 2: ```suggestion
- Offering a consistent way to write code across web and iOS, which doesn’t water down either platform.
```

> - Ensuring we were able to own our dependencies across the stacks

Reviewer 2: ```suggestion
- Ensuring we are able to meaningfully own our dependencies across the stacks.
```

> Meant that we took some technical and cultural debt specifically around the platform aspect of our native

Reviewer 2: ```suggestion
Means that we took on some technical and cultural debt, specifically around the platform aspect of our native
```

Until we discuss otherwise these principles still apply, so I changed them to be in present tense.

codebase.There are a few levers we can use to fix some of these issues:

> - Let iOS be a bit more webby, by using more JS instead of enforcing stricter platform standards

Reviewer 2: ```suggestion
- Let iOS be a bit more webby, by using more JS instead of enforcing stricter platform standards.
```

> - Use more of the React Native communities infra-structure, we generally don't use React Native JS dependencies

Reviewer 2: ```suggestion
- Use more of the React Native community’s infra-structure, we generally don't use React Native JS dependencies.
```

  These dependencies usually are cross-platform on Android and iOS which tends to mean making compromises

>   per-platform. Often we are forcing ourselves into extra work to ensure platform consistency

Reviewer 2: ```suggestion
  per-platform. Often we are forcing ourselves into extra work to ensure platform consistency.
```

> - Find a way to ensure clearer engineering and product ownership for iOS at Artsy

Reviewer 2: ```suggestion
- Find a way to ensure clearer engineering and product ownership for iOS at Artsy.
```

> - Have a lower mix of iOS infrastructure engineers to product engineers

Reviewer 2: Not sure if this means more or less iOS devs.

```suggestion
- Have a lower mix of iOS infrastructure engineers to product engineers.
```

Author: Good point, will re-phrase

How and if we'll tweak these levers will make for an interesting retrospective. Till then, I'd like to finish with
a showcase of some of the React Native work we shipped over the course of the last 4 months:

{% include epic_img.html url="/images/react-native-3/rn_3_1.jpg" title="" style="width:100%;" %}

{% include epic_img.html url="/images/react-native-3/rn_3_2.jpg" title="" style="width:100%;" %}

> We wouldn't have been able to ship this without React Native.

Reviewer 2: ❤️

and [iOS][fe-ios].

Maybe this is a small admission of failure to the idea of a purely de-silo'd team, but realistically while product
work across the two front-ends teams is consistent - the platform concerns just aren't.

## Keeping up with the Jobs'

As individual developers, it's much less important for us to keep up to date with the latest Apple developer news.
WWDCs and iOS releases just kinda happen, and we keep moving independently.

This is a side-effect to the openness of the JavaScript toolchain, and our ability to contribute and fix a lot of
our own problems at JavaScript level instead of needing to dive deeper into the native side to work on problems. We
still support iOS 9, and have very little incentive to bump it - new APIs from Apple just aren't that interesting
anymore.

Without a focused iOS team, it's not easy to pitch for iOS specific projects. This means that features like moving
to Apple's new app store review API or using iOS features which don't exist on web get spoken about in meetings,
but never started. This lack of a focused team makes it really hard to implement fixes to e.g. bad App Store

> reviews

Reviewer 3: Missing period
